File: src/main/java/com/burra/practice/leetcode/TopKFrequentElements.java
package com.burra.practice.leetcode;

import java.util.*;

public class TopKFrequentElements {

  public static int[] topKFrequent(int[] nums, int k) {
    HashMap<Integer, Integer> map = new HashMap<>();
    for (int num : nums) {
      map.put(num, map.getOrDefault(num, 0) + 1);
    }

    //get the max frequency
    int max = 0;
    for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
      max = Math.max(max, entry.getValue());
    }

    //initialize an array of ArrayList. index is frequency, value is list of numbers
    ArrayList<Integer>[] arr = (ArrayList<Integer>[]) new ArrayList[max + 1];
    for (int i = 1; i <= max; i++) {
      arr[i] = new ArrayList<>();
    }

    for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
      int count = entry.getValue();
      int number = entry.getKey();
      arr[count].add(number);
    }

    List<Integer> result = new ArrayList<>();

    //add most frequent numbers to result
    for (int j = max; j >= 1; j--) {
      if (!arr[j].isEmpty()) {
        for (int a : arr[j]) {
          result.add(a);
          //if size==k, stop
          if (result.size() == k) {
            return convtList(result);
          }
        }
      }
    }

    return convtList(result);
  }

  public static int[] convtList(List<Integer> l) {
    int[] a = new int[l.size()];
    int i = 0;
    for (Integer p : l) {
      a[i] = p;
      i++;
    }
    return a;
  }

  public static void main(String[] arg) {
    Arrays.stream(topKFrequent(new int[]{1, 1, 1, 2, 2, 3}, 2)).forEach(System.out::println);
  }
}

File: src/main/java/com/burra/practice/blind75/TopKFrequentElements.kt
package com.burra.practice.blind75

class TopKFrequentElements {
    /**
     * Bucket sort, or bin sort, is a sorting algorithm that works by distributing the elements of an array into a number of buckets. Each bucket is then sorted individually, either using a different sorting algorithm, or by recursively applying the bucket sorting algorithm.

    In this process we gonna follow 3 major steps :-

    Step - 1 :
    Create Frequency map:

    Step - 2 :
    Create Bucket List[]:

    Step - 3 :
    If we look at bucket arr we can see that most frequent elements are located at the end of arr
    and leat frequent elemnts at the begining
    Last step is to iterate from the end to the begining of the arr and add elements to result List
     */
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val frequencyMap = HashMap<Int, Int>()
        nums.forEach { frequencyMap[it] = frequencyMap.getOrDefault(it, 0)+1 }

        val bucketList = Array(nums.size + 1) { i -> arrayListOf<Int>() }

        frequencyMap.forEach { entry -> bucketList[entry.value].add(entry.key) }

        var index = 0
        val result = IntArray(k)
        for(i in bucketList.size-1 downTo 0){
            if(bucketList[i].isNotEmpty()) {
                for (j in bucketList[i]) {
                    result[index] = j
                    index++
                    if (index == k)
                        return result
                }
            }
        }
        return result
    }
}

fun main() {
    var topKFrequentElements = TopKFrequentElements()
    var result =  topKFrequentElements.topKFrequent(intArrayOf(1,1,1,2,2,3),2)
    println(result)
}